### PR TITLE
Create `Lrnr_glm_semiparametric`

### PR DESCRIPTION
**Pull request by Larsvanderlaan**
This learner provides a general way of fitting a semiparametric model using a baseline learner for the nonparametric component and glm for the parametric component. The use of family is analogous to glm and it can handle a parametric interaction with both binary and continuous variables.

### CODE REVIEW

**Comment by Larsvanderlaan** [comment]
Will add tests

**Comment by nhejazi** [comment]
Hi @Larsvanderlaan --- thanks for this contribution, it'll be a nice addition to the package. I'll defer a more thorough review until the unit tests have been added, but, in the meantime, I wanted to note a few code style changes that should be implemented.

1. Over the past couple of months, we migrated the documentation style (via a PR merged into `devel`) to use markdown, include examples, and more closely adhere to a common structure. For an example, consider [`Lrnr_haldensify`](https://github.com/tlverse/sl3/blob/devel/R/Lrnr_haldensify.R). Please update the documentation for this contribution to match the new format. Although incomplete, work on this front is continuing in https://github.com/tlverse/sl3/pull/355 and will be necessary for CRAN submission.
2. As a matter of convention, either try to adhere to the [`tidyverse` style guide](https://style.tidyverse.org/) (e.g., `snake_case` variable and function names, `FALSE` instead of `F`, etc.), either manually or by running `make style` prior to requesting a review. This helps with enforcing a consistent style and improving code readability for reviewers.

**Comment by Larsvanderlaan** [comment]
Hi @nhejazi, Thank you for the tips and references. I will document as is done in Lrnr_haldensify.

File: R/Lrnr_glm_semiparametric [hunk 1]
```diff
@@ -0,0 +1,251 @@
+#' Semiparametric Generalized Linear Models
```
Comment (nhejazi) [inline R/Lrnr_glm_semiparametric]: Please adhere to the [80-column rule](https://softwareengineering.stackexchange.com/questions/148677/why-is-80-characters-the-standard-limit-for-code-width) throughout, via manual or automatic line-breaking (you can enable this in RStudio, if that's the IDE you're using). `styler::style_pkg()` also enforces this automatically iirc.

File: R/Lrnr_glm_semiparametric [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+#'
+#' This learner provides fitting procedures for semiparametric generalized linear models using a user-given baseline learner and
+#' \code{\link[stats]{glm.fit}}. It supports models of the form `linkfun(E[Y|A,W]) = linkfun(E[Y|A=0,W]) + A * f(W)` where `A` is a binary or continuous interaction variable,
+#' and `f(W)` is a user-specified parametric function (e.g. `f(W) = model.matrix(formula_sp, W)`). The baseline function `E[Y|A=0,W]` is fit using a user-specified Learner (possibly pooled over values of `A` and then projected onto the semiparametric model).
```
Comment (nhejazi) [inline R/Lrnr_glm_semiparametric]: I believe your package referenced a paper by Mark, Sherri, Susan for some of this methodology. If that's correct, please add this as a reference appropriately, including addition of the bibTeX entry into the `REFERENCES.bib` file in `inst/`; for how to cite, see, for example, https://github.com/tlverse/sl3/blob/devel/R/Lrnr_glmnet.R#L8.

File: R/Lrnr_glm_semiparametric [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+#'   Note, if `append_interaction_matrix = TRUE`, the resulting estimator will be projected onto the semiparametric model using \code{glm.fit}.
+#'   If this is FALSE and \code{interaction_variable} is binary then the semiparametric model is learned by stratifying on \code{interaction_variable}.
+#'   Specifically, if FALSE, \code{lrnr_baseline} is used to estimate `E[Y|A=0,W]` by subsetting to only observations with `A = 0`.
+#'   In the binary case, setting `append_interaction_matrix = TRUE` allows one to pool the learning across treatment arms and allows additive models to perform well.
+#'   }
+#'   \item{\code{return_matrix_predictions = FALSE}}{Only used if \code{interaction_variable} is binary. Whether to return a matrix output with three columns being `E[Y|A=0,W], E[Y|A=1,W], E[Y|A,W]`.
+#'   }
+#'  \item{\code{...}}{Not used.
+#'  }
+#' }
```
Comment (nhejazi) [inline R/Lrnr_glm_semiparametric]: Switch these to use the new markdown syntax that we're using in the `devel` branch. Again, `Lrnr_glmnet` gives a fine example for how to do this. Basically, you can get ride of all the LaTeX macros, like `\describe{}`, `\item{}`, etc. Also see the comment above about following the 80-column rule.

File: R/Lrnr_glm_semiparametric [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+#'
+#' @examples
+#' library(glmnet)
+#' n <- 200
+#' W <- runif(n, -1, 1)
+#' A <- rbinom(n, 1, plogis(W))
+#' Y_continuous <- rnorm(n, mean = A + W, sd = 0.3)
+#' Y_binary <- rbinom(n, 1, plogis(A + W))
+#' Y_count <- rpois(n, exp(A + W))
+#' data <- data.table(W, A, Y_continuous, Y_binary, Y_count)
```
Comment (nhejazi) [inline R/Lrnr_glm_semiparametric]: It's great to see so many examples. Note, however, that they need to all run in under 5sec (per CRAN policy), so consider wrapping some of them in `\dontrun{...}` blocks if the timing is cumulatively beyond the 5sec limit.

**Review by nhejazi — CHANGES_REQUESTED** [review]
Overall, this looks good. Please address the comments about documentation, styling, and examples.

**Comment by rachaelvp** [comment]
Merged with current version of devel, updated documentation to adhere to 80-column rule, and replaced the ` symbols to denote code with `\code{}`. This learner is almost ready to be merged with devel. @Larsvanderlaan, we just need a test for the learner.

**Comment by rachaelvp** [comment]
@Larsvanderlaan, could you please look over the documentation in `Lrnr_glm_semiparametric`, including the description of the learner and the arguments, and add to the test if you think it's necessary? I tried to generalize the description of the functionality to prediction settings, so it's more inline with `sl3` formulation of just an outcome and covariates. Also, `W` was never defined but used in notation. I assumed this denoted all covariates in the task, excluding the interaction variable. 

@nhejazi, examples require loading `glmnet` package. Do you think this is OK?

**Comment by nhejazi** [comment]
if necessary, it's fine, but, if it's possible to rewrite the examples to just use `stats::glm` instead of `glmnet::glmnet` that would be better (to avoid adding `glmnet` as a hard dependency)

**Comment by Larsvanderlaan** [comment]
@rachaelvp Thank you very much for doing this. The updated documentation looks great. Your interpretation of `W` is correct. I also removed the dependency with glmnet (which I hopefully did without making any new errors).

**Comment by nhejazi** [comment]
Thanks @rachaelvp and @Larsvanderlaan for your work on this --- great to have this PR merged!

**Pull request merged**